smithy-utils/src/main/java/software/amazon/smithy/utils/CodeWriter.java

```diff
@@ -385,6 +401,33 @@
  * System.out.println(writer.toString());
  * // Outputs: "Names: Bob\n       Karen\n       Luis\n"
  * }</pre>
+ *
+ * <p>Alignment occurs either statically or dynamically based on the characters
+ * that come before interpolation. If all of the characters in the literal
+ * template that come before interpolation are spaces and tabs, then those
+ * characters are used when indenting newlines. Otherwise, the number of
+ * characters written as the template result that come before interpolation
+ * are used when indenting (this takes into account any interpolation that
+ * may precede block interpolation).
```

Comment:
I struggled to understand this for a while. I don't have any suggestions to make it clearer, so it may just be a problem with the reader; maybe more examples would help.